### ISSUE TITLE
deepDiffer causing lots of Array iteration logspew

### ISSUE THREAD

**Issue by ide** [facebook/react-native#868]
See:
```
Your code is broken! Do not iterate over arrays with for...in. See https://fburl.com/31944000 for more information.

    at Array.Object.defineProperty.Object.defineProperty.get [as __ARRAY_ENUMERATION_GUARD__] (Array.prototype.es6.js:26)
    at Object.deepDiffer [as diff] (deepDiffer.js:39)
    at diffRawProperties (diffRawProperties.js:54)
    at ReactIOSNativeComponent.Mixin.computeUpdatedProperties (ReactIOSNativeComponent.js:150)
    at ReactIOSNativeComponent.Mixin.receiveComponent (ReactIOSNativeComponent.js:189)
    at Object.ReactReconciler.receiveComponent (ReactReconciler.js:97)
    at ReactCompositeComponentMixin._updateRenderedComponent (ReactCompositeComponent.js:726)
    at ReactCompositeComponentMixin._performComponentUpdate (ReactCompositeComponent.js:705)
    at ReactCompositeComponentMixin.updateComponent (ReactCompositeComponent.js:625)
    at ReactPerf.measure.wrapper [as updateComponent] (ReactPerf.js:70)
```

**Comment by dvbportal** [comment]
For me it is unclear why for..in should be considered broken. (The link 'https://fburl.com/31944000' is gone)

I also refactored some of those loops into .forEach(function() {..., but there are some 'this'-related side effects.

**Comment by spicyj** [comment]
Sorry, the link goes to an FB-internal page:

![image](https://cloud.githubusercontent.com/assets/6820/7216407/6bcc4756-e5b3-11e4-9a47-a3b062862dcc.png)

You can pass `this` as a second argument to forEach or use an arrow function to preserve the context.

**Comment by dvbportal** [comment]
Thanks Ben, for clearing that up.  I see now, that it is a good idea to give that warning.

**Comment by josebalius** [comment]
@spicyj @dvbportal so guys is this still a bug or is there a proper way to do this? I am having the same error with ListView.

**Comment by spicyj** [comment]
Still an issue unfortunately. Should be fixed with the next sync though.

**Comment by josebalius** [comment]
@spicyj ah okay, it's a showstopper if you are using sections/rows, thanks for letting me know.

**Comment by soheil** [comment]
+1

**Comment by romanonthego** [comment]
+1
..and any updates? it's impossible to use many of libraries right now (baobab and typology for example) without monkey patching it.

**Comment by sahrens** [comment]
I landed a fix internally that should get sync'd out soon.

**Comment by ide** [comment]
@sahrens is your fix to make the polyfilled methods non-enumerable? That's the cleanest solution I had in mind since RN is used only in ES5+ environments.

**Comment by sahrens** [comment]
My fix just checks Array.isArray in deepDiffer and does the appropriate type of iteration.

**Comment by ide** [comment]
Got it - that fix seems right too. For the Array.prototype polyfill I think we want to use Object.defineProperty to add the polyfilled methods as non-enumerable properties. I imagine this was not done consciously for www so is it possible for www and RN to have different copies of the polyfill?

**Comment by vjeux** [comment]
It was done consciously for www because we want to have the same behavior with IE8 that doesn't support Object.defineProperty. If we were to use Object.defineProperty, we would break IE8 in a lot of terrible ways as most developers are not testing against IE8. But, you are right, it doesn't seem like a good way to polyfill it in React Native.

**Comment by spicyj** [comment]
Fixed in ab1efbd4c12a5792f55a1021903688d5aa864b15.
